File: lib/codegen.lua
CodeGen = class "CodeGen"

function CodeGen:initialize()
    self.buffer = {}
end

function CodeGen:adjust(n)
    if type(n) ~= "number" then error("expected number, got " .. type(n)) end
    table.insert(self.buffer, Insn(OpCode.ADJUST, n))
end

function CodeGen:select(n)
    if type(n) ~= "number" then error("expected number, got " .. type(n)) end
    table.insert(self.buffer, Insn(OpCode.SELECT, n))
end

function CodeGen:inc(n) self:adjust(math.abs(n or 1)) end
function CodeGen:dec(n) self:adjust(-math.abs(n or 1)) end

function CodeGen:right(n) self:select(math.abs(n or 1)) end
function CodeGen:left(n) self:select(-math.abs(n or 1)) end

function CodeGen:read(n) table.insert(self.buffer, Insn(OpCode.READ, n or 1)) end
function CodeGen:write(n) table.insert(self.buffer, Insn(OpCode.WRITE, n or 1)) end

function CodeGen:open()
    table.insert(self.buffer, Insn(OpCode.OPEN))
end

function CodeGen:close()
    table.insert(self.buffer, Insn(OpCode.CLOSE))
end

function CodeGen:set(x)
    if type(x) ~= "number" then error("expected number, got " .. type(x)) end
    table.insert(self.buffer, Insn(OpCode.SET, x))
end

function CodeGen:clear()
    self:set(0)
end

File: src/codegen.rs
use crate::{alloc, ir::*};

pub fn generate_brainfuck(ir: &IR) -> String {
    let mut result = String::new();

    let block_locations = alloc::layout_blocks(&ir.blocks);
    let mut pointer: i32 = 0;

    for insn in &ir.insns {
        match &insn.variant {
            InsnVariant::Adjust(n) => {
                if *n > 0 {
                    result.push_str("+".repeat(*n as usize).as_str());
                } else {
                    result.push_str("-".repeat(n.abs() as usize).as_str());
                }
            },
            InsnVariant::Select(n) => {
                pointer += n;
                if *n > 0 {
                    result.push_str(">".repeat(*n as usize).as_str());
                } else {
                    result.push_str("<".repeat(n.abs() as usize).as_str());
                }
            },
            InsnVariant::Read(n) => result.push_str(",".repeat(*n as usize).as_str()),
            InsnVariant::Write(n) => result.push_str(".".repeat(*n as usize).as_str()),
            InsnVariant::Open => result.push('['),
            InsnVariant::Close => result.push(']'),
            InsnVariant::Set(x) => {
                result.push_str("[-]");
                if *x != 0 { result.push_str("+".repeat(*x as usize).as_str()) }
            },
            InsnVariant::To(blkid) => {
                let blk_loc = block_locations[*blkid as usize];
                if blk_loc as i32 > pointer {
                    let delta = blk_loc as i32 - pointer;
                    result.push_str(">".repeat(delta as usize).as_str());
                    pointer += delta;
                } else if (blk_loc as i32) < pointer {
                    let delta = pointer - blk_loc as i32;
                    result.push_str("<".repeat(delta as usize).as_str());
                    pointer -= delta;
                }
            },
        }
    }

    result
}